activestorage/app/jobs/active_storage/create_variants_job.rb

```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+class ActiveStorage::CreateVariantsJob < ActiveStorage::BaseJob
+  queue_as { ActiveStorage.queues[:transform] }
+
+  discard_on ActiveRecord::RecordNotFound
+  retry_on ActiveStorage::IntegrityError, attempts: 10, wait: :polynomially_longer
+
+  def perform(blob, transformations:, generate:)
+    @blob = blob
+    @generate = generate
+
+    @blob.preview({}).processed if preview_image_needed?
+
+    transformations.each do |transformation|
+      ActiveStorage::TransformJob.public_send(perform_method, @blob, transformation)
+    end
+  end
+
+  private
+    def perform_method
+      (@generate == :immediate) ? :perform_now : :perform_later
```

Comment:
The job uses the `@generate` to determine whether it needs to perform the `TransformJob`s now or later. I couldn't find a way for the job itself to know whether it is being run with a `perform_now` or a `perform_later`, so I passed it explicitly to the job.